llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+			blockToIndexMap[bb] = decreasing_index;
+		}
+		blockQueue.emplace_back(data, this, subset);
+		decreasing_index--;
+	}
+	blockToIndexMap[start] = 0;	
+}
+
+void FunctionData::actualVisit(bool skipVisit = false)
+{
+	BasicBlockGroupData groupData(*this);
+	if (skipVisit)
+		groupData.visitAll();
+	else
+		groupData.recursiveVisit();
+}
+
+static void processFunction(const llvm::Function& F, ModuleData& moduleData)
+{
+	using namespace llvm;
+	//For each SCC (process in order)
+	//	if single point of entry: start from there, otherwise (no point of entry -> all unreachable / multiple ones -> all reachable)
+	//	execute from single point of entry (+ state that's intersection of possible path to it)
+	//	if (unconditional or decidable) -> follow path, otherwise
+	//		if immediate dominator is in the SCC, jump there(a refinement would be immediate dominator in SCC while invalidating all otherwise reachable SCC) or bail out
+	//	bail out:
+	//		sign as reachable all out-going edges from the SCC (keeping only the state pre-SCC) + all BB in the SCC
+
+	if (F.isDeclaration())
+		return;
+
+	FunctionData& data = moduleData.getFunctionData(F);
+	bool hasIndirectUseOrExternal = false;
+
+	if (F.getLinkage() != GlobalValue::InternalLinkage)
+		hasIndirectUseOrExternal = true;
+
+	for (const Use &U : F.uses())
+	{
+		if (hasIndirectUseOrExternal)
+			break;
+		const User *FU = U.getUser();
+		if (!isa<CallInst>(FU))
+		{
+			hasIndirectUseOrExternal = true;
+			break;
+		}
+		//Normally this pattern should handle also InvokeInst, but here support is not yet present 
+		const CallBase* CS = cast<CallBase>(FU);
+		if (CS->isCallee(&U))
+		{
+			data.visitCallBase(CS);	
+		}
+		else
+		{
+			hasIndirectUseOrExternal = true;
+		}
+	}
+	
+	if (hasIndirectUseOrExternal)
+	{
+		data.callEquivalentQueue.clear();
+		data.enqueVisitNoInfo();	
+	}
+
+	while (const FunctionData::VectorOfArgs* toBeVisited = data.getSomethingToVisit())
+	{
+		data.visitCallEquivalent(*toBeVisited);
+	}	
+}
+
+static bool modifyFunction(llvm::Function& F, ModuleData& moduleData)
+{
+	using namespace llvm;
+
+	if (F.isDeclaration())
+		return false;
+
+	bool changed = false;
+
+	FunctionData& data = moduleData.getFunctionData(F);
+	if (data.hasModifications(/*emitStats*/ false))
+	{
+		//Remove the edges that have never been taken
+		data.cleanupBB(F);
+		changed = true;
+	}
+
+	if (!data.callEquivalentQueue.empty())
+	{
+		//Check whether all call-sites have, for a given numArgs (without counting VAArgs), always the same Value
+		const uint32_t numArgs = data.callEquivalentQueue.front().size();
+
+		for (uint32_t i=0; i<numArgs; i++)
+		{
+			const llvm::Value* candidate = data.callEquivalentQueue[0][i];
+			bool equivalentArg = true;
+			for (uint32_t j=1; equivalentArg && j<data.callEquivalentQueue.size(); j++)
+			{
+				const llvm::Value* v = data.callEquivalentQueue[j][i];
+				if (candidate != v)
+					equivalentArg = false;
+			}
+
+			if (equivalentArg && candidate)
```

Comment:
Can you verify if this ever triggers? I would think that the rest of the pipeline would have done this forwarding already

Reply:
It triggers, but it's minor and muddies the waters, removed